stdlib/public/Darwin/Foundation/Data.swift

```diff
@@ -716,6 +716,14 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
             }
         }
 
+        @inlinable // This is @inlinable as tribially computable.
+        mutating func append(byte: UInt8) {
+            let count = self.count
+            assert(count + 1 <= MemoryLayout<Buffer>.size)
+            Swift.withUnsafeMutableBytes(of: &bytes) { $0[count] = byte }
```

Comment:
This looks the almost the same as the code that was causing the slowness of byte-by-byte appending of the sequence in the first place… `Swift.withUnsafeMutableBytes`. I might be mistaken about how the `withUnsafeMutableBytes` is compiled down to final assembler, but my hunch is this isn't the fastest way to do this… @atrick might have an opinion.

Reply:
I'm also confused reading this by the mixing of `count`, the explicit assert on `count + 1`, writing to `count` (not +1?), and direct modification of `length` (skipping of precondition check on `set count`)

Reply:
`bytes` here is a 7- or 15-byte tuple with guaranteed sequential layout. `withUnsafeMutableBytes(of:)` is the only way to write into an offset into the tuple's layout — this should incur no cost, and inline into the same code as `bytes.0`/`bytes.1`/`bytes.2`/etc. If it does not, we have a huge, huge problem here. (The alternative, BTW, is `switch`ing on `count` and assigning via `.0`/`.1`/`.2`, which really should not be necessary.)

As for the usage of `.count` over `.length`, let me annotate line by line:

```swift
let count = self.count // Necessary to stash locally to allow us to safely use this in the closure -- otherwise we risk conflicting mutable access to `self`
assert(count + 1 <= MemoryLayout<Buffer>.size) // Ensure that we've got space to append 1 byte to our statically-sized buffer
Swift.withUnsafeMutableBytes(of: &bytes) { $0[count] = byte /* append the byte */ }
self.length += 1 // Take care of the length; no need to go through count's precondition since we've already asserted this to be valid earlier
```

Reply:
I don’t think asserts and preconditions produce the same code under varying optimization levels. Shouldn’t this at least happened be using the same construct?

Benchmark has shown only 5x speed up with this code, less than ~10x I’d expect from buffer size, so I would definitely try the `switch(count)` approach. But I might be wrong here, as my `withUnsafeBytes` experience comes only from brief learning in #20552…

Reply:
OK, I [re-read on differences](https://stackoverflow.com/a/29673504/41307) between assert and precondition and I see you probably chose `assert` because it would get eliminated in `-O` build. That makes sense!

Reply:
I'm willing to give it a shot, but I would be very, very worried if we saw a 2x improvement via a `switch` over direct access with `withUnsafeMutableBytes(of:)`.

Reply:
Well, if that doesn't get much closer to `ExactCount` performance, we could try buffering into bigger array here. `ExactCount` is just dumping the sequence into raw memory with capacity that's known beforehand. This needs to perform more or less the same, our only concern here should be to not blow too much the memory during this. The problem with the original code:
````swift
Swift.withUnsafeBytes(of: element) {
                        _representation.append(contentsOf: $0)
                    }
````
is the huge amount of overhead you incur per-byte by lacking a direct single byte append on `_DataStore`. Is there another way you could write directly to the pre-allocated memory on `_DataSlice` while there is capacity?

Reply:
As expected, I see no meaningful difference here with

```diff
-            Swift.withUnsafeMutableBytes(of: &bytes) { $0[count] = byte }
+            switch count {
+            case 0: bytes.0 = byte
+            case 1: bytes.1 = byte
+            case 2: bytes.2 = byte
+            case 3: bytes.3 = byte
+            case 4: bytes.4 = byte
+            case 5: bytes.5 = byte
+#if arch(x86_64) || arch(arm64) || arch(s390x) || arch(powerpc64) || arch(powerpc64le)
+            case 6: bytes.6 = byte
+            case 7: bytes.7 = byte
+            case 8: bytes.8 = byte
+            case 9: bytes.9 = byte
+            case 10: bytes.10 = byte
+            case 11: bytes.11 = byte
+            case 12: bytes.12 = byte
+            case 13: bytes.13 = byte
+#endif
+            default: fatalError("Invalid InlineData.count value")
+            }
```

Results on my machine:

```diff
#,TEST,SAMPLES,MIN(μs),MAX(μs),MEAN(μs),SD(μs),MEDIAN(μs)
-116,Data.append.Sequence.ExactCount,10,125,128,127,1,127
-117,Data.append.Sequence.UnderestimatedCount,10,1167,1178,1171,4,1169
-121,Data.init.Sequence.ExactCount,10,108,111,109,1,109
-122,Data.init.Sequence.UnderestimatedCount,10,1226,1256,1241,11,1240
+116,Data.append.Sequence.ExactCount,10,121,128,124,2,123
+117,Data.append.Sequence.UnderestimatedCount,10,1165,1187,1172,6,1170
+121,Data.init.Sequence.ExactCount,10,107,110,108,1,108
+122,Data.init.Sequence.UnderestimatedCount,10,1205,1233,1213,8,1213
```

I can try increasing the size of the buffer by not using `InlineData`, but we're always going to hit tradeoffs here — by not being able to preallocate the space, we're always going to eat the cost of growing the data eventually. (Keep in mind we also don't necessarily have a `_DataStorage` to write into directly — if the sequence is small enough, we're specifically trying to avoid allocation here.)

Reply:
Hmmm… that interesting.

I've been looking at the assembler this produces over at https://swift.godbolt.org/z/2b7zNL
`Swift.withUnsafeMutableBytes` produces the cleanest code. Switch produces few extra jumps and `fatalError` complicates the epilogue quite a bit… but it apparently makes no difference.

Reply:
As expected. :)

Also, for what it's worth, I've tested switching out the `InlineData` with larger and larger `ContiguousArray`s (even up to 512 bytes), and as expected, there is no improvement in performance. The overhead of calling `next()` over and over again (as opposed to bulk filling out a preallocated buffer) is quite large. I'm trying to profile locally to see exactly what is most expensive here, but it's slow-going with other stuff going on.

Reply:
Why can't you just add `append(UInt8)` like you now have on `InlineData` to the `_DataSlice` and forward to it from `InlineSlice` and `LargeSlice`? `_DataSlice` already has amortized growth logic with `ensureUniqueBufferReference(growingTo:clear)`… I think all we need to do is avoid the dance around `UnsafeRawPointer` that the single currently available `append`, right?

Reply:
(I'm assuming you mean `_DataStorage` here, since we don't have `_DataSlice`)
We can add single-byte append throughout and try, but again, I don't think this is the slowness we're hitting — the issue I'm seeing is the overhead of doing this byte-by-byte, which is something we simply can't get around.

I'll give it a shot either way.

Reply:
I'm curious where does the [`_copyContents`](https://github.com/apple/swift/blob/master/stdlib/public/core/Sequence.swift#L1093) get its performance advantage then… it is also calling [`it.next()`](https://github.com/apple/swift/blob/master/stdlib/public/core/Sequence.swift#L1099) in a loop.

Looking again at the code path that `ExactCount` takes, why aren't you allowing `_copyContents` to write directly into `_DataStorage`, but are using an extra buffer. Isn't that avoidable, and a whole point of using this method? It seems to be using 2x the memory.

Reply:
Yeah, as unfortunately expected:

```diff
diff --git a/stdlib/public/Darwin/Foundation/Data.swift b/stdlib/public/Darwin/Foundation/Data.swift
index 699d334669..b07486de14 100644
--- a/stdlib/public/Darwin/Foundation/Data.swift
+++ b/stdlib/public/Darwin/Foundation/Data.swift
@@ -283,6 +283,19 @@ internal final class _DataStorage {
         }
         _length = newLength
     }
+
+    @usableFromInline // This is not @inlinable as it is a non-trivial, non-generic function.
+    func append(_ byte: UInt8) {
+        let origLength = _length
+        let newLength = origLength + 1
+        if _capacity < newLength || _bytes == nil {
+            ensureUniqueBufferReference(growingTo: newLength, clear: false)
+        }
+
+        _bytes!.storeBytes(of: byte, toByteOffset: _length, as: UInt8.self)
+        _length = newLength
+    }
+
     
     @inlinable // This is @inlinable as it does not escape the _DataStorage boundary layer.
     func append(_ bytes: UnsafeRawPointer, length: Int) {
@@ -717,7 +730,7 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
         }
 
         @inlinable // This is @inlinable as tribially computable.
-        mutating func append(byte: UInt8) {
+        mutating func append(_ byte: UInt8) {
             let count = self.count
             assert(count + 1 <= MemoryLayout<Buffer>.size)
             Swift.withUnsafeMutableBytes(of: &bytes) { $0[count] = byte }
@@ -973,6 +986,14 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
             return try storage.withUnsafeMutableBytes(in: range, apply: apply)
         }
 
+        @inlinable // This is @inlinable as reasonably small.
+        mutating func append(_ byte: UInt8) {
+            assert(endIndex + 1 < HalfInt.max)
+            ensureUniqueReference()
+            storage.append(byte)
+            slice = slice.lowerBound..<HalfInt(Int(slice.upperBound) + 1)
+        }
+
         @inlinable // This is @inlinable as reasonably small.
         mutating func append(contentsOf buffer: UnsafeRawBufferPointer) {
             assert(endIndex + buffer.count < HalfInt.max)
@@ -1182,6 +1203,13 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
             return try storage.withUnsafeMutableBytes(in: range, apply: apply)
         }
 
+        @inlinable // This is @inlinable as reasonably small.
+        mutating func append(_ byte: UInt8) {
+            ensureUniqueReference()
+            storage.append(byte)
+            slice.range = slice.range.lowerBound..<(slice.range.upperBound + 1)
+        }
+
         @inlinable // This is @inlinable as reasonably small.
         mutating func append(contentsOf buffer: UnsafeRawBufferPointer) {
             ensureUniqueReference()
@@ -1525,6 +1553,40 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
             }
         }
 
+        @inlinable // This is @inlinable as reasonably small.
+        mutating func append(_ byte: UInt8) {
+            switch self {
+            case .empty:
+                var inline = InlineData()
+                inline.append(byte)
+                self = .inline(inline)
+            case .inline(var inline):
+                if inline.count < inline.capacity {
+                    defer { self = .inline(inline) }
+                    self = .empty
+                    inline.append(byte)
+                } else {
+                    var slice = InlineSlice(inline)
+                    slice.append(byte)
+                    self = .slice(slice)
+                }
+            case .slice(var slice):
+                if slice.count < slice.capacity {
+                    defer { self = .slice(slice) }
+                    self = .empty
+                    slice.append(byte)
+                } else {
+                    var large = LargeSlice(slice)
+                    large.append(byte)
+                    self = .large(large)
+                }
+            case .large(var large):
+                defer { self = .large(large) }
+                self = .empty
+                large.append(byte)
+            }
+        }
+
         @inlinable // This is @inlinable as reasonably small.
         mutating func append(contentsOf buffer: UnsafeRawBufferPointer) {
             switch self {
@@ -2078,21 +2140,9 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
                 // Copy the contents of buffer...
                 _representation = _Representation(UnsafeRawBufferPointer(start: base, count: endIndex))
 
-                // ... and append the rest byte-wise, buffering through an InlineData.
-                var buffer = InlineData()
+                // ... and append the rest byte-wise.
                 while let element = iter.next() {
-                    if buffer.count < buffer.capacity {
-                        buffer.append(byte: element)
-                    } else {
-                        buffer.withUnsafeBytes { _representation.append(contentsOf: $0) }
-                        buffer.count = 0
-                    }
-                }
-
-                // If we've still got bytes left in the buffer (i.e. the loop ended before we filled up the buffer and cleared it out), append them.
-                if buffer.count > 0 {
-                    buffer.withUnsafeBytes { _representation.append(contentsOf: $0) }
-                    buffer.count = 0
+                    _representation.append(element)
                 }
             }
         }
@@ -2372,21 +2422,9 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
             // Copy the contents of the buffer...
             _representation.append(contentsOf: UnsafeRawBufferPointer(start: base, count: endIndex))
 
-            // ... and append the rest byte-wise, buffering through an InlineData.
-            var buffer = InlineData()
+            // ... and append the rest byte-wise.
             while let element = iter.next() {
-                if buffer.count < buffer.capacity {
-                    buffer.append(byte: element)
-                } else {
-                    buffer.withUnsafeBytes { _representation.append(contentsOf: $0) }
-                    buffer.count = 0
-                }
-            }
-
-            // If we've still got bytes left in the buffer (i.e. the loop ended before we filled up the buffer and cleared it out), append them.
-            if buffer.count > 0 {
-                buffer.withUnsafeBytes { _representation.append(contentsOf: $0) }
-                buffer.count = 0
+                _representation.append(element)
             }
         }
     }
```

produces

```diff
#,TEST,SAMPLES,MIN(μs),MAX(μs),MEAN(μs),SD(μs),MEDIAN(μs)
-116,Data.append.Sequence.ExactCount,10,125,128,127,1,127
-117,Data.append.Sequence.UnderestimatedCount,10,1167,1178,1171,4,1169
-121,Data.init.Sequence.ExactCount,10,108,111,109,1,109
-122,Data.init.Sequence.UnderestimatedCount,10,1226,1256,1241,11,1240
+116,Data.append.Sequence.ExactCount,10,120,124,122,1,121
+117,Data.append.Sequence.UnderestimatedCount,10,1149,1158,1153,2,1153
+121,Data.init.Sequence.ExactCount,10,109,111,110,0,110
+122,Data.init.Sequence.UnderestimatedCount,10,1203,1228,1213,9,1209
```

Reply:
@palimondo I'm not 100% sure, to be honest — a lot of this gets inlined beyond easy recognition. At least part of the reason is that `_withStackOrHeapBuffer` is able to `alloca` a large stack buffer for us to `copyContents` into — and in this specific test (with these size restrictions: `1033 + 809 = 1842 < 2048`), we can very quickly grab _all_ the elements out of the sequence locally first (in the exact count case), and then initialize a `_Representation` wholesale from that, without regrowing on the heap at all, ever.

In the unknown case, we can't allocate the large stack buffer, so we grow and regrow on the heap, while also pulling elements out one-by-one.

The whole point of going through `_withStackOrHeapBuffer`, BTW, is to both avoid that initial allocation, and to avoid allocating at _all_ if we don't have to. For data that fits in `InlineData`, we don't _ever_ want to create a `_DataStorage`.

Reply:
The other `append` on `_DataStorage` is `@inlineable`, this could be the reason for no perf. change:
````swift
@usableFromInline // This is not @inlinable as it is a non-trivial, non-generic function.
    func append(_ byte: UInt8) {
````

Both are calling the `ensureUniqueBufferReference(growingTo...` which is `@usableFromInline`, so that's not it… the other thing in there is `_bytes.storeBytes` which I don't see anywhere, but that should't be that complicated, right?

Reply:
`_bytes.storeBytes` boils down to a direct byte assignment into memory; I'll play with making `append(_:)` `@inlinable` despite escaping the `_DataStorage` boundary, but I don't think it will matter much either :(

Reply:
@atrick Any thoughts about all of this? Any tricks we might be able to pull here to improve this case?

Reply:
Marking `_DataStorage.append(_:)` `@inlinable` instead of `@usableFromInline` regresses performance way back:

```diff
#,TEST,SAMPLES,MIN(μs),MAX(μs),MEAN(μs),SD(μs),MEDIAN(μs)
-116,Data.append.Sequence.ExactCount,10,125,128,127,1,127
-117,Data.append.Sequence.UnderestimatedCount,10,1167,1178,1171,4,1169
-121,Data.init.Sequence.ExactCount,10,108,111,109,1,109
-122,Data.init.Sequence.UnderestimatedCount,10,1226,1256,1241,11,1240
+116,Data.append.Sequence.ExactCount,10,120,127,123,2,122
+117,Data.append.Sequence.UnderestimatedCount,10,3961,4024,4003,18,4007
+121,Data.init.Sequence.ExactCount,10,109,114,110,2,109
+122,Data.init.Sequence.UnderestimatedCount,10,5398,5576,5471,62,5449
```

Reply:
We have a problem with what we're really measuring here… I'll have a new benchmark PR later today. And we'll probably need some stdlib and perf folks to weigh in.

Reply:
Do you mind elaborating on the gravity of this? At this point, we need to get these changes in ASAP — will this be blocking or can we merge?

Reply:
Non-blocking. I just re-tested with sequence that only differs in how it handles the `underestimatedCount` and it has different de-optimization characteristics than the two repeated sequence types we've been using, so we should change those two benchmarks and deal with the (de-)optimizations later.

Reply:
In this case I'm going to re-smoke-test and merge and we can follow-up on this separately.

Reply:
Resolving this convo. Benchmark refinement follow-up in #21848.